src/modules/developer_tools.json

```diff
@@ -0,0 +1,250 @@
+                        "name": "Default Result",
+                        "value": null
+                    }
+                }
+            ]
+        },
+        {
+            "name": "pair",
+            "summary": "Configures Firebolt to accept incoming Firebolt JSON RPC messages from the developer tool identified by the given developer tool ID",
+            "tags": [],
+            "params": [
+                {
+                    "name": "developerToolId",
+                    "summary": "The ID of the DeveloperTool to which to connect",
+                    "required": true,
+                    "schema": {
+                        "type": "string",
+                        "format": "uuid"
+                    }
+                }
+            ],
+            "result": {
+                "name": "result",
+                "schema": {
+                    "const": null
+                }
+            },
+            "examples": [
+                {
+                    "name": "Default Example",
+                    "params": [
+                        {
+                            "name": "developerToolId",
+                            "value": "ff6340fb-8576-494c-97c6-e60e9a4eeba9"
+                        }
+                    ],
+                    "result": {
+                        "name": "Default Result",
+                        "value": null
+                    }
+                }
+            ]
+        },
+        {
+            "name": "unpair",
+            "summary": "Disconnect Firebolt as a web socket client from the web socket that corresponds to the given developer tool ID",
+            "tags": [],
+            "params": [
+                {
+                    "name": "developerToolId",
+                    "summary": "The ID of the DeveloperTool from which to disconnect",
+                    "required": true,
+                    "schema": {
+                        "type": "string",
+                        "format": "uuid"
+                    }
+                }
+            ],
+            "result": {
+                "name": "result",
+                "schema": {
+                    "const": null
+                }
+            },
+            "examples": [
+                {
+                    "name": "Default Example",
+                    "params": [
+                        {
+                            "name": "developerToolId",
+                            "value": "ff6340fb-8576-494c-97c6-e60e9a4eeba9"
+                        }
+                    ],
+                    "result": {
+                        "name": "Default Result",
+                        "value": null
+                    }
+                }
+            ]
+        },
+        {
+            "name": "onDeveloperToolUp",
+            "tags": [
+                {
+                    "name": "event"
+                }
+            ],
+            "summary": "Listen to the developerToolUp event",
```

Comment:
Does this event get fired, but only after you call `find`?

Wifi returns a list from it's scan API, since all APIs are asynchronous, it "works."

I'm not sure I like it that way... I think both Wifi & DNS-SD could return more and more results over time, right? Like when you open a wifi menu you get some results, the more fill in over time. I assume DNS-SD is the same.

It'd be nice to have a similar API pattern for both of these.

Maybe something like a new method tag called `scanning` that generates an API that looks like this:

```javascript
const scanner = DeveloperTools.scan( results => {
  // results is an array of DeveloperTool objects, e.g. { name: 'foo', id: 'bar' }
  results.forEach( result => {
    let item = document.createElement('li')
    item.innerHTML = result.name
    myToolList.appendChild(item)
  }
})

cancelButton.onclick = () => {
  scanner.cancel()
}
```

Reply:
I don't think what we're doing with WiFi is the right approach. There, just like here, the events are asynchronous and come in over time. That's partly why "find" is long-running... a tool can be started after the find is invoked and then show up in the list a few milliseconds/seconds later. The DNS-SD implementations and docs I've seen all are event-based as tools are found ('up' in the DNS-SD parlance) or go away ('down').

Reply:
Not sure quite how the "scan" thing above would work... how long would it wait to get results? The whole point is that the query is long-running and events (available/unavailable) fire as appropriate while the find is running.